lib/mutgen.c

```diff
@@ -187,93 +184,494 @@ mutation_model_check_validity(mutation_model_t *self)
+    site->ancestral_state_length = 0;
+    return 0;
+}
+
+static int
+slim_mutator_transition(mutation_model_t *self, gsl_rng *MSP_UNUSED(rng),
+    const char *parent_allele, tsk_size_t parent_allele_length,
+    const char *parent_metadata, tsk_size_t parent_metadata_length, mutation_t *mutation)
+{
+    int ret = 0;
+    slim_mutator_t *params = &self->params.slim_mutator;
+    char *buff = NULL;
+    int len;
+    /* The maximum number of digits for a signed 64 bit integer (including
+     * the leading "-") */
+    const size_t max_digits = 20;
+    /* We allow for a potential comma to separate previous elements in the
+     * list. We allow for a possible comma to separate the previous element
+     * in the list, as well as a NULL byte added by snprintf. We don't bother
+     * trying to alloc the exact number of bytes needed. */
+    const size_t alloc_size = parent_allele_length + max_digits + 2;
+    const char *sep = parent_allele_length == 0 ? "" : ",";
+
+    /* Append to derived_state */
+    buff = tsk_blkalloc_get(&params->allocator, alloc_size);
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    len = snprintf(buff, alloc_size, "%.*s%s%" PRId64, parent_allele_length,
+        parent_allele, sep, params->next_mutation_id);
+    if (len < 0) {
+        /* Technically this can happen. Returning TSK_ERR_IO should result
+         * in Python code checking errno. */
+        ret = TSK_ERR_IO;
+        goto out;
+    }
+    assert(len < (int) alloc_size);
+    if (params->next_mutation_id == INT64_MAX) {
+        ret = MSP_ERR_MUTATION_ID_OVERFLOW;
+        goto out;
+    }
+    params->next_mutation_id++;
+    mutation->derived_state = buff;
+    mutation->derived_state_length = (tsk_size_t) len;
+
+    /* Append to metadata */
+    buff = tsk_blkalloc_get(
+        &params->allocator, parent_metadata_length + SLIM_MUTATION_METADATA_SIZE);
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    memcpy(buff, parent_metadata, parent_metadata_length);
+    copy_slim_mutation_metadata(params, buff + parent_metadata_length);
+
+    mutation->metadata = buff;
+    mutation->metadata_length
+        = (tsk_size_t)(parent_metadata_length + SLIM_MUTATION_METADATA_SIZE);
+out:
+    return ret;
+}
+
+static int
+slim_mutator_free(mutation_model_t *self)
+{
+    slim_mutator_t params = self->params.slim_mutator;
+    tsk_blkalloc_free(&params.allocator);
+    return 0;
+}
+
+/***********************
+ * Infinite alleles mutation model */
+
+static void
+infinite_alleles_print_state(mutation_model_t *self, FILE *out)
+{
+    infinite_alleles_t params = self->params.infinite_alleles;
+    fprintf(out, "infinite alleles:\n");
+    fprintf(out, "\tstart_allele:%" PRIu64 "\n", params.start_allele);
+}
+
+/* The maximum number of digits for an unsigned 64 bit integer is 20
+ * and one byte for the NULL terminator. */
+#define MAX_UINT_BUFF_SIZE 21
```

Comment:
we could use this above (`max_digits` in slim_transition_allele) as well

Reply:
I thought about this, but it's a bit awkward as we'd have to have another implementation of make_allele as we're using uint64 in one and int64 in the other. The reason for having ``uint64`` here, btw, is so that we have well defined overflow semantics. Unsigned ints are guaranteed to wrap around to zero, whereas signed ints are undefined. Having overflow seemed simpler than having to test for it, but we could totally do it same way as the slim model too.

Anyway, it didn't seem worth complicating the code more, as the slim code is pretty water tight and reasonably simple now. The extra few bytes spent per allele is trivial.